### PR TITLE
Adds Razer Huntsman V2 Support

### LINKED ISSUE #1620: Support for Razer Huntsman V2 [1532:026c]
### Device Info

* Name: Razer Huntsman v2
* Product Number: RZ03-03930400-R3U1
* Model Number: RZ03-0393

### Device Information
```
Bus 001 Device 005: ID 1532:026c Razer USA, Ltd Razer Huntsman V2
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x026c 
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer Huntsman V2
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```
```
lrwxrwxrwx 1 root root  10 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-event-if01 -> ../event12
lrwxrwxrwx 1 root root  10 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-event-if04 -> ../event14
lrwxrwxrwx 1 root root  10 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-event-kbd -> ../event10
lrwxrwxrwx 1 root root  10 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-if01-event-kbd -> ../event11
lrwxrwxrwx 1 root root  10 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-if02-event-mouse -> ../event13
lrwxrwxrwx 1 root root   9 Sep 26 14:09 usb-Razer_Razer_Huntsman_V2-if02-mouse -> ../mouse2
```

### Packet Captures

<Please see here as we probably need packet captures>
>

Can do if needed .

### Synapse Screenshots

Can do if needed.

### PR DESCRIPTION
**Pull request by nightsky30**
Adds Razer Huntsman V2 Support

Fixes #1620

Note:  Looks like the Huntsman V2: Analog could also benefit from these sorts of changes/additions

Not Working/Implemented:
- [x] Macros ? (No dedicated M keys... I can either leave the macro stuff in or remove)

Works:
 - [x] Breath Single
 - [x] Breath Dual
 - [x] Breath Random
 - [x] Starlight Single
 - [x] Starlight Dual
 - [x] Starlight Random
 - [x] Ripple
 - [x] Ripple Random
 - [x] Reactive
 - [x] Wave
 - [x] None
 - [x] Spectrum
 - [x] Static
 - [x] Brightness

**EDIT: Moved additional functionality and fixes to a separate PR**

~~Added Functionality:~~
- [x] ~~Poll Rate (did get packets)~~
  - ~~It seems to be displaying the same behavior and same values as other devices using `razer_chroma_misc_set_polling_rate2`~~
  - ~~Up to 8000 Hz~~
  - ~~Seems Razer added a 250 Hz poll rate.~~
- [x] ~~Added 250 Hz Poll Rate~~
- [x] ~~Keyswitch Optimization (did get packets)~~
  - ~~Typing vs Gaming~~
  - ~~Implemented as a toggle of state~~
  - ~~The 1st command (set of two commands sent each time) seems to be toggling the increased debounce delay and faster triggering for the modes.~~
  - ~~The 2nd command (set of two commands sent each time) is a near copycat that seems to be increasing the number of parameters by 1, using the first parameter as the VARSTORE, and shifting the other parameter indexes by 1.~~
  - ~~Keyswitch optimization state was verified saved between reboots using D-Feet and Wireshark.~~

~~Working (with fix):~~
- [x] ~~Game Mode LED~~
  - ~~Seems it works with the Huntsman V2 if the device files are created along with all the other files on the Mouse Protocol interface~~
  - ~~Tested my changes with both the Huntsman V2 and Blackwidow Chroma V2 which uses one of the additional Keyboard Protocol interfaces~~
```
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 1441, resource id: 23079522, major code: 40 (TranslateCoords), minor code: 0
libpng warning: iCCP: known incorrect sRGB profile
Traceback (most recent call last):
  File ’/home/user/Documents/Projects/github/polychromatic/polychromatic/pylib/backends/openrazer.py’, line 1013, in set_device_state
    rdevice.game_mode_led = option_data
  File ’/usr/lib/python3/dist-packages/openrazer/client/devices/keyboard.py’, line 34, in game_mode_led
    self._dbus_interfaces[’game_mode_led’].setGameMode(True)
  File ’/usr/lib/python3/dist-packages/dbus/proxies.py’, line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File ’/usr/lib/python3/dist-packages/dbus/connection.py’, line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Python.PermissionError: Traceback (most recent call last):
  File ’/usr/lib/python3/dist-packages/dbus/service.py’, line 711, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File ’/usr/lib/python3/dist-packages/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py’, line 85, in set_game_mode
    open(super_file, ’wb’).write(b’\x01’)
PermissionError: [Errno 13] Permission denied: ’/sys/devices/pci0000:00/0000:00:14.0/usb1/1-13/1-13:1.3/0003:1532:026C.0010/key_super’
```
```
/sys/devices/pci0000:00/0000:00:14.0/usb1/1-13/1-13:1.0/0003:1532:026C.000D$ ls -al
total 0
drwxr-xr-x 5 root plugdev    0 Nov 15 17:59 .
drwxr-xr-x 5 root root       0 Nov 15 17:59 ..
-r--r--r-- 1 root plugdev 4096 Nov 15 17:59 country
lrwxrwxrwx 1 root plugdev    0 Nov 15 17:59 driver -> ../../../../../../../bus/hid/drivers/razerkbd
drwxr-xr-x 3 root plugdev    0 Nov 15 17:59 hidraw
drwxr-xr-x 3 root plugdev    0 Nov 15 17:59 input
-rw-rw---- 1 root plugdev 4096 Nov 15 18:02 key_alt_f4
-rw-rw---- 1 root plugdev 4096 Nov 15 18:02 key_alt_tab
-rw-rw---- 1 root plugdev 4096 Nov 15 18:02 key_super
-r--r--r-- 1 root plugdev 4096 Nov 15 17:59 modalias
drwxr-xr-x 2 root plugdev    0 Nov 15 17:59 power
-r--r--r-- 1 root plugdev 4096 Nov 15 17:59 report_descriptor
lrwxrwxrwx 1 root plugdev    0 Nov 15 17:59 subsystem -> ../../../../../../../bus/hid
-rw-r--r-- 1 root plugdev 4096 Nov 15 17:59 uevent
```

### Device Model Info

* Name: Razer Huntsman V2
* Product Number: RZ03-03930400-R3U1
* Model Number: RZ03-0393

### Device Information

<details>
<summary>
ls -lah /dev/input/by-id/
</summary>

lrwxrwxrwx 1 root root  10 Nov 15 17:59 usb-Razer_Razer_Huntsman_V2-event-if04 -> ../event18
lrwxrwxrwx 1 root root  10 Nov 15 17:59 usb-Razer_Razer_Huntsman_V2-event-kbd -> ../event14
lrwxrwxrwx 1 root root  10 Nov 15 17:59 usb-Razer_Razer_Huntsman_V2-if01-event-kbd -> ../event16
lrwxrwxrwx 1 root root  10 Nov 15 17:59 usb-Razer_Razer_Huntsman_V2-if02-event-mouse -> ../event17
lrwxrwxrwx 1 root root   9 Nov 15 17:59 usb-Razer_Razer_Huntsman_V2-if02-mouse -> ../mouse4

</details>
<details>
<summary>
lsusb -d 1532: -v
</summary>

Bus 001 Device 008: ID 1532:026c Razer USA, Ltd Razer Huntsman V2
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x026c 
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer Huntsman V2
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Qualifier (for other device speed):
  bLength                10
  bDescriptorType         6
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  bNumConfigurations      0
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)
</details>

- Supports Transaction ID 0x1f (maybe 0x3f as well, hadn't tried this)
- Game Mode / LED seems to be handled on the mouse protocol interface, unlike older keyboards
- Has 8k polling rate support
- Has some kind of keyswitch optimization

### Screenshots
Screenshot_2021-11-15_17-36-41
Screenshot_2021-11-15_17-44-10

### Packets
[Huntsman_V2.zip](https://github.com/openrazer/openrazer/files/7542452/Huntsman_V2.zip)

### CODE REVIEW

**Comment by nightsky30** [comment]
@z3ntu @lah7 Fixed the Game Mode issue with the Huntsman V2. Not sure if it's the best solution. I'm open to your thoughts on the fix.

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -434,8 +435,17 @@ static ssize_t razer_attr_write_mode_game(struct device *dev, struct device_attr
 {
     struct usb_interface *intf = to_usb_interface(dev->parent);
     struct usb_device *usb_dev = interface_to_usbdev(intf);
+    struct razer_report report = {0};
     unsigned char enabled = (unsigned char)simple_strtoul(buf, NULL, 10);
-    struct razer_report report = razer_chroma_standard_set_led_state(VARSTORE, GAME_LED, enabled);
+
+    switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
+        report = razer_chroma_standard_set_led_state(NOSTORE, GAME_LED, enabled);
```
Comment (nightsky30) [inline driver/razerkbd_driver.c:527]: The packet capture indicated Synapse was not using VARSTORE. I could just modify the transaction ID and continue using VARSTORE with the report if you think that best.

**Comment by nightsky30** [comment]
Also, looks like the same issue with the Game Mode might affect the Blackwidow V3 (Full Size) #1728 , and it got missed in the discussions around building/testing in the PR: https://github.com/openrazer/openrazer/pull/1572#issuecomment-926585972

**Comment by ThomChroma** [comment]
Do you want to add support for the TKL version as well or should I open a new issue? The outputs are the same, as are the options available on Synapse. Is it just a new entry to add for that ID?

### Device Info

* Name: Razer Huntsman V2 Tenkeyless
* Product Number: RZ03-03940600-R3F1
* Model Number: RZ03-0394

### Device Information
```
Bus 001 Device 004: ID 1532:026b Razer USA, Ltd Razer Huntsman V2 Tenkeyless
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x026b
  bcdDevice            1.01
  iManufacturer           1 Razer
  iProduct                2 Razer Huntsman V2 Tenkeyless
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     177
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      0
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.00
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-if01 -> ../event17
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-if04 -> ../event19
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if01-event-kbd -> ../event7
lrwxrwxrwx 1 root root  10 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if02-event-mouse -> ../event18
lrwxrwxrwx 1 root root   9 24 nov.  14:16 usb-Razer_Razer_Huntsman_V2_Tenkeyless-if02-mouse -> ../mouse3
```

**Comment by nightsky30** [comment]
@ThomChroma I'd suggest you open a new PR. Glad you were able to make use of this one though to confirm stuff works with the TKL version :smile:

**Comment by ThomChroma** [comment]
Actually I didn't because I'm new to OpenRazer and have no idea how to add the TKL version :smiling_face_with_tear: I was just guessing it will just works :tm: as they seem to be identical.

**Comment by nightsky30** [comment]
> Actually I didn't because I'm new to OpenRazer and have no idea how to add the TKL version smiling_face_with_tear I was just guessing it will just works tm as they seem to be identical.

Ah, I see. Yes, you most definitely will need a different PR. I also would modify the wording of your comment in the issue you created. Reading it, makes it sound as if this PR fixes it or adds support for the TKL version when it does not. Maybe I'm misinterpreting. That would involve more changes and separate testing. This current PR at best provides some required elements and a template for which you can work with when adding support for the TKL version.

**Comment by ThomChroma** [comment]
Indeed that wasn't clear at all. I meant to say that this PR was adding support for the full-sized version, to point out that there is already support for their features, so that someone doesn't do it again and just builds on what you did if that's possible :+1:

**Comment by r4v** [comment]
Hey,
Is there any chance to fix conflicts and merge this to the project? I'm waiting two months to have support my keyboard in openrazer. Thanks in advance

**Comment by z3ntu** [comment]
@nightsky30 Would you be open to ripping out the new functionality (keyswitch optimization and whatnot) for now. I'm a bit reluctant of adding new APIs at the moment :wink: Just moving them to a new PR would be fine for me so basic support could be merged.

**Comment by saibotk** [comment]
As soon, as we get this in, i think, im going to also adjust my PR #1754 to rebase on the latest changes here and also only add basic support first.
So we could quickly support both new devices at least with the basic functionality.

**Comment by Chris2000SP** [comment]
Ok, how do i test this? I have lsusb 1532:026c on arch aur/openrazer-meta 3.2.0-1 installed. [Polychromatic](https://github.com/polychromatic/polychromatic) shows a unknown device "? 1532:026C".

EDIT: i found the wiki. I try the build instruktions.

**Comment by Chris2000SP** [comment]
I have build and installed it but i get nothing more than the driver with your repo installed and i get the same as before. "? 1532:026C" If i click on that device there says "Unrecognized Device"

EDIT: oh my bad, it is in other branch.

**Comment by Chris2000SP** [comment]
Works for me! I get "Brightness" "Effect" "Game Mode" "Macros" "Key Mapping" as possible config options.

:+1:

**Comment by r4v** [comment]
> "Macros"

What? Huntsman V2 does not have dedicated M1-M5 keys, so macro recording does not work

**Comment by Chris2000SP** [comment]
It has Macros over the "fn" + "F9" keys. I don't know how to use it. I look in Manual.

**Comment by Chris2000SP** [comment]
@r4v you're right. It only works with that non existing Keys. Also Polychromatic don't support it though.

EDIT: In Manual says "4. Press the desired key where you want to save your macro". Does this mean i can save it on any key?

**Comment by Chris2000SP** [comment]
:heart: ! Syncronous with my Basilisk v2.

**Comment by Chris2000SP** [comment]
i dunno if that is the fault of this patch or not but if i set synchronized my huntsman and basilisk it drifts of in time. I have to set it again to sync.

EDIT: found issue to this: https://github.com/polychromatic/polychromatic/issues/389

**Comment by iamKunal** [comment]
Hi @nightsky30 when can we plan to have this merged. I'm actually awaiting #1754 by @saibotk for the TKL :grimacing:

**Comment by nightsky30** [comment]
> Hi @nightsky30 when can we plan to have this merged. I'm actually awaiting #1754 by @saibotk for the TKL grimacing

@iamKunal That is not my decision to make. I simply contribute and wait patiently. If you desire it sooner, you might consider [building](https://github.com/openrazer/openrazer/wiki/Building-a-package) that particular branch.

**Comment by dominik1999w** [comment]
> I have build and installed it but i get nothing more than the driver with your repo installed and i get the same as before. "? 1532:026C" If i click on that device there says "Unrecognized Device"
> 
> EDIT: oh my bad, it is in other branch.

@Chris2000SP I have the same issue when building from this branch. Could you explain what you did?

**Review by z3ntu — APPROVED** [review]
LGTM, thanks a lot!

CI is a bit weird today so I had to trigger it manually: https://builds.sr.ht/~z3ntu/job/731968#setup-139

**Comment by nightsky30** [comment]
> LGTM, thanks a lot!
> 
> CI is a bit weird today so I had to trigger it manually: https://builds.sr.ht/~z3ntu/job/731968#setup-139

Yeah, this PR CI was stuck for quite a while. No worries :smile:

**Comment by kierun** [comment]
Any idea when the next release will happen? I run Fedora and kinda like to keep to packages instead of having to re-compile all the time. Thank you.

**Comment by z3ntu** [comment]
@kierun Quite soon, planning for the next couple of days.

**Comment by 0pLuS0** [comment]
> @kierun Quite soon, planning for the next couple of days.

Any word?

THANKS

**Comment by kierun** [comment]
> @0pLuS0 
> Any word?
> 
> THANKS

It has been released. There's a bug with media keys but otherwise, it works fine.

**Comment by kierun** [comment]
https://github.com/openrazer/openrazer/issues/1800#event-6833157173

Marked as closed and fixed, now waiting for next release.

**Pull request merged**